cli/command/trust/signer_add.go

```diff
@@ -0,0 +1,141 @@
+package trust
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/command/image"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/cli/opts"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/tuf/data"
+	tufutils "github.com/docker/notary/tuf/utils"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+type signerAddOptions struct {
+	keys   opts.ListOpts
+	signer string
+	images []string
+}
+
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	var options signerAddOptions
+	cmd := &cobra.Command{
+		Use:   "add OPTIONS NAME IMAGE [IMAGE...] ",
+		Short: "Add a signer",
+		Args:  cli.RequiresMinArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			options.images = args[1:]
+			return addSigner(dockerCli, options)
+		},
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.Var(&options.keys, "key", "Path to the signer's public key file(s)")
+	return cmd
+}
+
+var validSignerName = regexp.MustCompile(`^[a-z0-9][a-z0-9\_\-]*$`).MatchString
+
+func addSigner(cli command.Cli, options signerAddOptions) error {
+	signerName := options.signer
+	if !validSignerName(signerName) {
+		return fmt.Errorf("signer name \"%s\" must start with lowercase alphanumeric characters and can include \"-\" or \"_\" after the first character", signerName)
+	}
+	if signerName == "releases" {
+		return fmt.Errorf("releases is a reserved keyword, please use a different signer name")
+	}
+
+	if options.keys.Len() == 0 {
+		return fmt.Errorf("path to a public key must be provided using the `--key` flag")
+	}
+	signerPubKeys, err := ingestPublicKeys(options.keys.GetAll())
+	if err != nil {
+		return err
+	}
+	var errImages []string
+	for _, imageName := range options.images {
+		fmt.Fprintf(cli.Out(), "Adding signer \"%s\" to %s...\n", signerName, imageName)
+		if err := addSignerToImage(cli, signerName, imageName, signerPubKeys); err != nil {
```

Comment:
Not a blocker, but noticed that we're not consistent in requiring a `:tag` or not;

```bash
$ docker trust sign thajeztah/signingtest
No tag specified for thajeztah/signingtest
```

But

```bash
$ docker trust signer add bob thajeztah/signingtest --key=/foo/baz/bob.pub

Adding signer "bob" to thajeztah/signingtest...
Enter passphrase for repository key with ID 05cab95:
Successfully added signer: bob to thajeztah/signingtest
```

Reply:
for signing a tag is required to specify an image.  For `signer add` and `signer remove` we silently parse out any tags because they operate over the full image repo - we did this for convenience, but we could error out explicitly if you prefer?

Reply:
Ah, right, so

```
docker trust signer add bob thajeztah/signingtest:one thajeztah/signingtest:two
```

is basically doing

```
docker trust signer add bob thajeztah/signingtest:*
```

So `docker trust signer add` adds a signer to a _repository_, not an _image_ ? I know "repository" is a bit overloaded, but perhaps we need to use that here (again: could use some input on that one)

Reply:
yes it is, great catch! Notary operates a single trusted collection per repository. I'll update usage and argnames accordingly